Make artifact transforms more efficient

oehme opened the pull request:
Only create a build operation if we don't already
have a cached value in memory. Otherwise we end up
flooding our build operation infrastructure with
synchronization requests and also create lots of
garbage descriptor objects.

oehme commented:
This makes the Studio Sync mockup 7% [faster](https://builds.gradle.org/viewLog.html?buildId=9328969&tab=buildResultsDiv&buildTypeId=Gradle_Util_Performance_AdHocPerformanceScenarioLinux).

jjohannes commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultArtifactTransforms.java:
```diff
@@ -172,7 +171,11 @@ public void collectBuildDependencies(BuildDependenciesVisitor visitor) {
+                    operation.run(null);
+                } else {
+                    actions.add(operation);
+                }
```
:o: Would be good to document this condition with some test coverage.

jjohannes commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/VariantAttributeMatchingCache.java:
```diff
@@ -117,4 +107,41 @@ private boolean matchAttributes(AttributeContainerInternal actual, AttributeCont
+                return true;
+            }
+            return false;
+        }
```
:o: I think we do not have any test coverage for `hasCachedResult()`. Since it needs to be in sync with what `transform()` does we should add some. Otherwise it would be easy to break. Since `transform()` is always called anyway; independent of the `hasCachedResult ()` result.
It could be unit test coverage for the two implementations of  `transform()` / `hasCachedResult()` and/or some coverage for the build operation handling in `ConsumerProvidedResolvedVariant` (see other comment).

oehme commented:
You're right, this was a bit rushed. I split out all those inner classes into their own files, so we can write more targeted tests. I started with tests for the caching behavior. I also added a bit of documentation to the new transformer interface.
